### DISCUSSION TITLE
Finding Dusk Time

### DISCUSSION THREAD

**Discussion by Strangename99** [micropython/micropython#9642] [Using MicroPython]
Python can import Astral to find various astronpmical times, such as dusk time for your location.   Is there an equivalent with micropython?  If not is there another way to do this without resorting to tables?

**Comment by scruss** [comment]
If you need this much more accurately than ±5 minutes or so, MicroPython's maths handling may not be accurate enough for you.

I don't know of a module that explicitly calculates dusk (any of the civil, nautical or astronomical twilights), but Dan J Perron put a simple module [here](https://forums.raspberrypi.com/viewtopic.php?t=340080) that calculates sunrise and sunset. The start of twilight doesn't vary a huge amount from sunset over the year, but the amount depends on your latitude.

Note that MicroPython's time zone handling is not very advanced. Typically, localtime will always return the same as gmtime. We're a very small system, after all.

**Reply by Strangename99** [reply]
Thanks .  I can cope with the (in)accuracy.  Will take a look.

**Comment by bulletmark** [comment]
I have a [small application](https://github.com/bulletmark/wcscheduler) that runs on Python on a Raspberry Pi that does this. I just fetch the sunrise/sunset times from https://sunrise-sunset.org/api.

**Reply by peterhinch** [reply]
An excellent source of astronomical algorithms with explanations of their operation is "Astronomy on the Personal Computer" by Oliver Montenbruck and Thomas-Pfleger.

Before MP I used their algorithms in a clock that displays sunrise and set, moon rise and set, and moon phase (the book goes much further than this).

**Reply by Strangename99** [reply]
This looks similar to Astral, albeit a cut-down version.   Might work well for me.  I'll have a play.

**Reply by scruss** [reply]
> An excellent source of astronomical algorithms with explanations of their operation is "Astronomy on the Personal Computer" by Oliver Montenbruck and Thomas-Pfleger.

similarly, *Astronomical Algorithms* by Jean Meeus is exceptionally complete. Unfortunately, it's astronomically expensive.

This page has a complete breakdown of how it's calculated: [Sunrise equation - Wikipedia](https://en.wikipedia.org/wiki/Sunrise_equation)

**Comment by DavesCodeMusings** [comment]
In addition to the Wikipedia article mentioned, there is another page: https://sciencing.com/calculate-suns-declination-6904335.html that lays out step by step how to do the calculations. I used the information there to create a [sunrise sunset calculator](https://github.com/DavesCodeMusings/SolarEdgeAPI/blob/main/daylight.js) that might fit what you're looking for. Unfortunately, it's written in JavaScript, not Python. But, it's heavily commented and should be able to serve as an example of the math involved if you wanted to write your own.
